internal/telemetry/tracker.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"encoding/json"
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/mongodb/mongocli/internal/config"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+const (
+	cacheFilename           = "telemetry"
+	dirPermissions          = 0700
+	filePermissions         = 0600
+	defaultMaxCacheFileSize = 100_000_000 // 100MB
+)
+
+type tracker struct {
+	fs               afero.Fs
+	maxCacheFileSize int64
+	cacheDir         string
+}
+
+func newTracker() (*tracker, error) {
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return nil, err
+	}
+
+	cacheDir = filepath.Join(cacheDir, config.ToolName)
+
+	return &tracker{
+		fs:               afero.NewOsFs(),
+		maxCacheFileSize: defaultMaxCacheFileSize,
+		cacheDir:         cacheDir,
+	}, nil
+}
+
+func (t *tracker) track(cmd *cobra.Command, e error) error {
+	options := []eventOpt{withCommandPath(cmd), withDuration(cmd), withFlags(cmd), withProfile(), withVersion(), withOS(), withAuthMethod(), withService(), withProjectID(cmd), withOrgID(cmd), withTerminal(), withInstaller(t.fs)}
```

Comment:
[q] Why are you creating different functions to set the properties if they are always set? Why not have one function `newProperties` that adds all the properties for an event?

Reply:
The code was originally together linter was complaining about it getting too big and it was had to follow.

Since we already use the option pattern in the CLI it felt the right way to modularize the code.

Reply:
Thanks!